src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,6 +104,95 @@ def pileupSizeTask(self):
+        - add new ruleIds to pileup document
+
+        :param rseExpression: string, Rucio rse expression, see
+        https://rucio.cern.ch/documentation/started/concepts/rse_expressions/
+        :return: None
+        """
+        spec = {'containerFraction': {"$gt": 0}}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
+            fraction = doc['containerFraction']
+
+            # check previous fraction value and proceed only if has changed
+            lastFraction = 0
+            transition = doc.get('transition', [])
+            if transition:
+                lastFraction = transition[-1]['containerFraction']
+            if lastFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # by default we assume it is decreasing
+            pname = doc['customName']
+            blockNames = self.rucioClient.getBlocksInContainer(pname)
+
+            # but if it increasing
+            if lastFraction < fraction:
+                pname = doc['pileupName']
+                blockNames += self.rucioClient.getBlocksInContainer(pname)
+
+            # get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+            portion = blockNames[:math.ceil(fraction * len(blockNames))]
+
+            self.logger.info("For pileup document '%s' take portion '%s'", doc, blockNames)
+
+            # create new container DID in Rucio
+            for blockName in portion:
+                self.rucioClient.createContainer(blockName, scope=self.customRucioScope)
+
+            # add new transition record
+            transition = doc['transition']
+            tdoc = transition[-1]
+            trec = {'customDID': tdoc['customDID'], 'updateTime': gmtimeSeconds(),
+                    'containerFraction': tdoc['containerFraction'], 'DN': tdoc['DN']}
+            transition.append(trec)
+            doc['transition'] = transition
+
+            # we call attachDIDs Rucio wrapper API with our set of DIDs
+            for rse in doc['currentRSEs']:
+                # superDID upper structure level (can be a container or block. If attaching blocks,
+                # then it's a container name
+                superDID = portion[0].split('#')[0]
+                self.rucioClient.attachDIDs(rse, superDID, portion, scope=self.customRucioScope)
+
+            # create new rules for custom DID using provide by rseExpression
+            newRules = self.rucioClient.createReplicationRules(portion, rseExpression)
```

Comment:
As mentioned before, this `rseExpression` needs to be taken from the pileup document itself.

In the scenario where we increase the pileup fraction, we will likely not have all blocks on Disk. Which makes me wonder: what happens when we attach those Tape blocks saying that they are available in RSE Disk_A and Disk_B? Maybe a question for Eric V and/or Rucio experts.

Reply:
And we need to create a replication rule for each RSE. If we pass in an RSE expression with multiple RSEs, my understanding is that it will create a single rule id, which is not what we want.

Reply:
ok